include/gsl/gsl_assert

```diff
@@ -113,19 +113,20 @@ namespace details
 
 #endif
 
-    [[noreturn]] inline void terminate() noexcept
-    {
 #if defined(GSL_MSVC_USE_STL_NOEXCEPTION_WORKAROUND)
-        (*gsl::details::get_terminate_handler())();
+    inline void terminate() noexcept { (*gsl::details::get_terminate_handler())(); }
```

Comment:
what is the reason for removing the [[noreturn]] here?

Reply:
Clang (v7.0.0) sees this as returning `terminate_handler` and throws the warning `-Winvalid-noreturn` here.

I don't understand this workaround for the STL's exceptions _(top of `namespace details` ~line 100)_.
But, if this was not intended (and it is not a bug in clang's implementation), than it could be changed there.

If you want to maintain the `[[noreturn]]` or the issue is with Clang, than I could disable the warning.
I chose not do this, since I couldn't figure out the validity of the warning. Considering that when a code or compiler update warrants adding `[[noreturn]]` again, a new warning will show up.

Reply:
I suppose this is because clang_cl gives a warning on invalind noreturn. This `terminate()` should never return, and we shouldn't have difference in API based on whether we use exceptions or not. Can we instead detect that we are compiling with clang and use `__attribute((noreturn))` in the function pointer typedef? msvc does not have a problem with this code, and gcc does not run on it.

`typedef void (__attribute((noreturn))  *func )();`

Reply:
Is clan_cl using MS STL? if yes, it has the same problem of std::terminate not terminating as msvc does, hence the workaround:)

Reply:
Yes, clang-cl uses the MS STL.

Reply:
It took me a while to get to some understanding.
And indeed, inconsistent use of `[[noreturn]]` in an interface would be an issue.

Considering using `__attribute__((noreturn))`:
It becomes part of the type `terminate_handler` and a requirement for every instantiation of the type and any custom terminate handler. Since the attribute is not the same as `[[noreturn]]` this creates a separate interface just for clang on Windows.
See: Farwaykorse/GSL commit: 9a171ce14a68c157c6db834fbcd038678eafda96

To me, this looks like an even worse situation.

Suppressing the warning locally seems the best option.
There seem to be no usage or compilation issues, when ignoring the warning, and it is limited to the workaround.
For end-users the noreturn interface properties should be maintained, and without warnings.

If I'm wrong here and there would be a better solution, please let me know. Currently I don't see it.